xdsl/pattern_rewriter.py

```diff
@@ -793,68 +792,129 @@ class PatternRewriteWalker:
+        if self.apply_recursively:
+            self._worklist.push(op)
+
+    def _handle_operation_replacement(
+        self, op: Operation, new_results: Sequence[SSAValue | None]
+    ) -> None:
+        """Handle replacement of an operation."""
+        if self.apply_recursively:
+            for result in op.results:
+                for user in result.uses:
+                    self._worklist.push(user.operation)
+
+    def _get_rewriter_listener(self) -> PatternRewriterListener:
+        """
+        Get the listener that will be passed to the rewriter.
+        It will take care of adding operations to the worklist, and calling the
+        listener passed as configuration to the walker.
+        """
+        return PatternRewriterListener(
+            operation_insertion_handler=(
+                self.listener.operation_insertion_handler
+                + [self._handle_operation_insertion]
+            ),
+            operation_removal_handler=(
+                self.listener.operation_removal_handler
+                + [self._handle_operation_removal]
+            ),
+            operation_modification_handler=(
+                self.listener.operation_modification_handler
+                + [self._handle_operation_modification]
+            ),
+            operation_replacement_handler=(
+                self.listener.operation_replacement_handler
+                + [self._handle_operation_replacement]
+            ),
+            block_creation_handler=(self.listener.block_creation_handler),
+        )
+
+    def rewrite_module(self, module: ModuleOp) -> None:
         """
-        for region in op.regions:
-            blocks = reversed(region.blocks) if self.walk_reverse else region.blocks
-            for block in blocks:
-                iter_op = block.last_op if self.walk_reverse else block.first_op
-                while iter_op is not None:
-                    iter_op = self._rewrite_op(iter_op)
+        Rewrite operations nested in the given operation by repeatedly applying the
+        pattern.
+        """
+        self.rewrite_op(module)
+
+    def rewrite_op(self, op: Operation) -> None:
+        """
+        Rewrite operations nested in the given operation by repeatedly applying the
+        pattern.
+        """
+        pattern_listener = self._get_rewriter_listener()
+
+        self._populate_worklist(op)
+        op_was_modified = self._process_worklist(pattern_listener)
+
+        if not self.apply_recursively:
+            return
+
+        while op_was_modified:
+            self._populate_worklist(op)
+            op_was_modified = self._process_worklist(pattern_listener)
+
+    def _populate_worklist(self, op: Operation) -> None:
+        """Populate the worklist with all nested operations."""
+        # We walk in reverse order since we use a stack for our worklist.
+        for sub_op in op.walk(
+            reverse=not self.walk_reverse, region_first=not self.walk_regions_first
```

Comment:
That looks like we are not quite consistent with naming. How can it be that we need a not here?

Reply:
The comment is there for that ;)
You need to walk in reverse order because you have a stack (LIFO). The reason we have a stack is to mirror MLIR, which I believe they made the right choice to be honest.
Naming wise, it does make sense, since `walk` can be omitted in the `walk` method IMO.